File: apps/ios/BitOS/DesignSystem/HexIdentity.swift
import SwiftUI

/**
 * Hex identity system (unified feature spec §2.5/§4, APP-022): the BitOS
 * sovereign-identity motif. `HexShape` is the flat-top hexagon clip shared
 * by avatars, badges and brand tiles (web `.hex-clip` / Flutter `HexShape`
 * parity); `HexAvatarView` is the deterministic pubkey avatar used by every
 * surface until remote profile pictures load.
 */

/// Pure geometry + identity-color derivation (testable without UI).
enum HexIdentity {
    /// Flat-top hexagon vertices inscribed in `rect`
    /// (CSS `polygon(25% 6.7%, 75% 6.7%, 100% 50%, 75% 93.3%, 25% 93.3%, 0% 50%)`
    /// parity with the web `.hex-clip` / Flutter `hexPoints`: a *regular*
    /// flat-top hexagon touching the left/right edges with a 6.7% vertical
    /// inset — not one stretched to fill the square).
    static func vertices(in rect: CGRect) -> [CGPoint] {
        [
            CGPoint(x: rect.minX + rect.width * 0.25, y: rect.minY + rect.height * 0.067),
            CGPoint(x: rect.minX + rect.width * 0.75, y: rect.minY + rect.height * 0.067),
            CGPoint(x: rect.maxX, y: rect.midY),
            CGPoint(x: rect.minX + rect.width * 0.75, y: rect.maxY - rect.height * 0.067),
            CGPoint(x: rect.minX + rect.width * 0.25, y: rect.maxY - rect.height * 0.067),
            CGPoint(x: rect.minX, y: rect.midY),
        ]
    }

    /// Deterministic pubkey → avatar gradient. Stable for the same pubkey
    /// across launches, distinct between keys.
    static func avatarGradient(for pubkey: String) -> (start: Color, end: Color) {
        let seed = UInt32(pubkey.prefix(8), radix: 16) ?? 0
        let hue = Double(seed % 360) / 360
        let hueEnd = (hue + 40.0 / 360.0).truncatingRemainder(dividingBy: 1)
        return (
            Color(hue: hue, saturation: 0.65, brightness: 0.85),
            Color(hue: hueEnd, saturation: 0.7, brightness: 0.55)
        )
    }
}

/// The hexagon clip used for avatars, identity chips and brand tiles.
struct HexShape: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        path.addLines(HexIdentity.vertices(in: rect))
        path.closeSubpath()
        return path
    }
}

/// Hexagonal identicon avatar: deterministic gradient + initials, with an
/// optional remote picture that replaces the fallback once it loads.
struct HexAvatarView: View {
    let pubkey: String
    var size: CGFloat = 40
    var imageURL: URL?
    var label: String?
    var hasLightning: Bool = false

    init(pubkey: String, size: CGFloat = 40, imageURL: URL? = nil, label: String? = nil, hasLightning: Bool = false) {
        self.pubkey = pubkey
        self.size = size
        self.imageURL = imageURL
        self.label = label
        self.hasLightning = hasLightning
    }

    private var initials: some View {
        Text(avatarInitials(label ?? pubkey))
            .font(.system(size: size / 3, weight: .bold))
            .foregroundStyle(.white)
    }

    var body: some View {
        let colors = HexIdentity.avatarGradient(for: pubkey)
        ZStack(alignment: .bottomTrailing) {
            ZStack {
                LinearGradient(
                    colors: [colors.start, colors.end],
                    startPoint: .topLeading,
                    endPoint: .bottomTrailing
                )
                if let imageURL {
                    AsyncImage(url: imageURL) { image in
                        image.resizable().scaledToFill()
                    } placeholder: {
                        initials
                    }
                } else {
                    initials
                }
            }
            .frame(width: size, height: size)
            .clipShape(HexShape())
            if hasLightning {
                AppIcons.image(for: AppIcons.zap)
                    .font(.system(size: min(max(size * 0.18, 6), 10), weight: .bold))
                    .foregroundStyle(.white)
                    .frame(width: min(max(size * 0.36, 10), 18), height: min(max(size * 0.36, 10), 18))
                    .background(Circle().fill(LinearGradient(colors: [Color(red: 1, green: 0.71, blue: 0.11), Color(red: 0.97, green: 0.58, blue: 0.10)], startPoint: .topLeading, endPoint: .bottomTrailing)))
                    .overlay(Circle().stroke(.white, lineWidth: 1))
            }
        }
        .frame(width: size, height: size)
        .accessibilityLabel("Avatar \(avatarInitials(label ?? pubkey))")
    }
}

/// Profile metadata is untrusted. Avatar fetches use only bounded HTTPS URLs
/// and never load remote SVG documents; callers retain the hex fallback.
func safeProfilePictureURL(_ raw: String?) -> URL? {
    guard let raw = raw?.trimmingCharacters(in: .whitespacesAndNewlines),
          raw.count <= 512,
          let url = URL(string: raw),
          url.scheme?.lowercased() == "https",
          url.host != nil else { return nil }
    guard !url.path.lowercased().hasSuffix(".svg") else { return nil }
    return url
}

private func avatarInitials(_ label: String) -> String {
    let words = label.split(whereSeparator: \.isWhitespace)
    switch words.count {
    case 0: return "?"
    case 1: return String(words[0].prefix(2)).uppercased()
    default: return "\(words[0].prefix(1))\(words[words.count - 1].prefix(1))".uppercased()
    }
}

/// Small hexagonal tile with a symbol (settings tiles, brand marks).
/// Decorative: label at the usage site.
struct HexIcon: View {
    let systemName: String
    var size: CGFloat = 40
    var background: Color = BitOSTheme.surfaceElevated
    var foreground: Color = BitOSTheme.accent

    var body: some View {
        ZStack {
            HexShape().fill(background)
            AppIcons.image(for: systemName)
                .font(.system(size: size * 0.42, weight: .semibold))
                .foregroundStyle(foreground)
        }
        .frame(width: size, height: size)
        .accessibilityHidden(true)
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/ui/components/HexIdentity.kt
package space.bitos.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Bolt
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Outline
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.Shape
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

/**
 * Hex identity system (unified feature spec §2.5/§4, APP-022): the BitOS
 * sovereign-identity motif. `HexShape` is the flat-top hexagon clip shared
 * by avatars, badges and brand tiles (web `.hex-clip` / Flutter `HexShape`
 * / iOS `HexShape` parity); `HexAvatar` is the deterministic pubkey avatar
 * used by every surface until remote profile pictures load.
 */

/** Pure geometry + identity-color derivation (JVM-testable, no UI). */
object HexIdentity {

    /** Flat-top hexagon vertices inscribed in the box
     * (CSS `polygon(25% 6.7%, 75% 6.7%, 100% 50%, 75% 93.3%, 25% 93.3%, 0% 50%)`
     * parity with the web `.hex-clip` / Flutter `hexPoints`: a *regular*
     * flat-top hexagon touching the left/right edges with a 6.7% vertical
     * inset — not one stretched to fill the square). */
    fun vertices(width: Float, height: Float): List<Offset> = listOf(
        Offset(width * 0.25f, height * 0.067f),
        Offset(width * 0.75f, height * 0.067f),
        Offset(width, height * 0.5f),
        Offset(width * 0.75f, height * 0.933f),
        Offset(width * 0.25f, height * 0.933f),
        Offset(0f, height * 0.5f),
    )

    data class IdentityColors(val start: Color, val end: Color)

    /** Deterministic pubkey → avatar gradient. Stable for the same pubkey
     * across launches, distinct between keys. */
    fun identityColors(pubkey: String): IdentityColors {
        val seed = pubkey.take(8).toLongOrNull(16) ?: 0L
        val hue = (seed % 360) / 360f
        val hueEnd = (hue + 40f / 360f) % 1f
        return IdentityColors(hsv(hue, 0.65f, 0.85f), hsv(hueEnd, 0.7f, 0.55f))
    }

    /** HSV (0..1 ranges) to Compose color. */
    fun hsv(h: Float, s: Float, v: Float): Color {
        val i = (h * 6).toInt()
        val f = h * 6 - i
        val p = v * (1 - s)
        val q = v * (1 - f * s)
        val t = v * (1 - (1 - f) * s)
        return when (i % 6) {
            0 -> Color(v, t, p)
            1 -> Color(q, v, p)
            2 -> Color(p, v, t)
            3 -> Color(p, q, v)
            4 -> Color(t, p, v)
            else -> Color(v, p, q)
        }
    }
}

/** Flat-top hexagon clip (BitOS identity motif). */
class HexShape : Shape {
    override fun createOutline(
        size: Size,
        layoutDirection: LayoutDirection,
        density: Density,
    ): Outline {
        val v = HexIdentity.vertices(size.width, size.height)
        val path = Path().apply {
            moveTo(v[0].x, v[0].y)
            for (i in 1 until v.size) lineTo(v[i].x, v[i].y)
            close()
        }
        return Outline.Generic(path)
    }
}

/** Hexagonal identicon avatar: deterministic gradient + initials. Remote
 * profile pictures load through the media pipeline later; the identicon is
 * the stable fallback so feed rows never shift layout. */
@Composable
fun HexAvatar(
    pubkey: String,
    modifier: Modifier = Modifier,
    size: Int = 40,
    label: String? = null,
    hasLightning: Boolean = false,
) {
    val colors = HexIdentity.identityColors(pubkey)
    val initials = avatarInitials(label ?: pubkey)
    Box(
        modifier = modifier
            .size(size.dp)
            .semantics { contentDescription = "Avatar $initials" },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .size(size.dp)
                .clip(HexShape())
                .background(Brush.linearGradient(listOf(colors.start, colors.end))),
            contentAlignment = Alignment.Center,
        ) {
            Text(initials, color = Color.White, fontSize = (size / 3).sp, fontWeight = FontWeight.W700)
        }
        if (hasLightning) {
            Icon(
                Icons.Rounded.Bolt,
                contentDescription = "Lightning enabled",
                tint = Color.White,
                modifier = Modifier
                    .align(Alignment.BottomEnd)
                    .size((size * 0.36f).coerceIn(10f, 18f).dp)
                    .clip(androidx.compose.foundation.shape.CircleShape)
                    .background(Brush.linearGradient(listOf(Color(0xFFFFB51B), Color(0xFFF7931A))))
                    .border(1.dp, Color.White, androidx.compose.foundation.shape.CircleShape)
                    .padding(2.dp),
            )
        }
    }
}

private fun avatarInitials(label: String): String {
    val words = label.trim().split(Regex("\\s+")).filter(String::isNotEmpty)
    return when (words.size) {
        0 -> "?"
        1 -> words.first().take(2).uppercase()
        else -> "${words.first().first()}${words.last().first()}".uppercase()
    }
}
